### PR TITLE
9 file format checks

### LINKED ISSUE #9: File format checks
Files provided are not checked against the format they should be and so the error thrown isn't useful.

### PR DESCRIPTION
**Pull request by BenOnTheBoard**
This branch contains format checking for instances of SM, HR and SPA-S instances; dictionary and file readers are both included.
Includes a set of error classes for use in future file checking and pre-processing.

Major changes:
- Readers now raise informative errors when they detect a misformatted input.
- Pre-processing (inc. deleting unacceptable pairs and creating rank dictionaries) is now entirely handled by the smPreferenceInstance, hrPreferenceInstance and spaPreferenceInstance classes, which leaves to the readers only the work which is actually reading. (SRP)
- Overall the readability of the code is massively improved.

### CODE REVIEW

File: src/algmatch/stableMatchings/hospitalResidentsProblem/hrPreferenceInstance.py [hunk 1]
```diff
@@ -19,4 +23,41 @@ def _load_from_file(self, filename: str) -> None:
...
+                
+    def clean_unacceptable_pairs(self) -> None:
+        for r, h in product(self.residents, self.hospitals):
+            if r not in self.hospitals[h]["list"] or h not in self.residents[r]["list"]:
+                try:
+                    self.residents[r]["list"].remove(h)
+                except ValueError:
+                    pass
+                try:
+                    self.residents[h]["list"].remove(r)
```
Comment (VaradK62442) [inline src/algmatch/stableMatchings/hospitalResidentsProblem/hrPreferenceInstance.py]: this line raises a `KeyError` when I run `hrVerifyCorrectness.py`. If I change the except to `KeyError`, there is another error raised when actually constructing the stable matching. Not sure what the aim of this code is?

File: src/algmatch/studentProjectAllocation.py [hunk 2]
```diff
@@ -40,5 +41,6 @@ def get_stable_matching(self) -> dict | None:
         :return: dict, the stable matching.
         """
         self.spa.run()
-        if not self.spa.blocking_pair: return self.spa.stable_matching
+        if not self.spa.is_stable:
```
Comment (VaradK62442) [inline src/algmatch/studentProjectAllocation.py]:
```suggestion
        if self.spa.is_stable:
```

**Comment by VaradK62442** [comment]
error handling is very nice, apart from the above two things, im happy to merge after making sure it all works fine :)

File: tests/SPATests/spaVerifyCorrectness.py [hunk 3]
```diff
@@ -33,13 +33,13 @@ def verify_instance(self):
...
-        l = StudentProjectAllocation(filename=filename, optimisedSide="lecturer")
+        L = StudentProjectAllocation(filename=filename, optimisedSide="lecturer")
 
         e.choose()
         s_stable_matching = s.get_stable_matching()
-        l_stable_matching = l.get_stable_matching()
+        L_stable_matching = L.get_stable_matching()
 
-        return l_stable_matching == e.all_matchings[-1] and s_stable_matching == e.all_matchings[0]
+        return L_stable_matching == e.all_matchings[-1] and s_stable_matching == e.all_matchings[0]
```
Comment (VaradK62442) [inline tests/SPATests/spaVerifyCorrectness.py]:
```suggestion
        return L_stable_matching['student_sided'] == e.all_matchings[-1] and s_stable_matching['student_sided'] == e.all_matchings[0]
```

File: src/algmatch/stableMatchings/hospitalResidentsProblem/hrPreferenceInstance.py (hunk 1 above)
Comment (BenOnTheBoard) [reply src/algmatch/stableMatchings/hospitalResidentsProblem/hrPreferenceInstance.py]: I think that should say self.hospitals[h] and is just a typo.
ValueError is raised if you try to remove an element from an array which is not present in it.

**Comment by BenOnTheBoard** [comment]
Ok I can now successfully run all the examples provided in the HR.
I also tried SM, HR and SPA-S with the existing test code (with a few different sizes) and they cleared 100% of instances.

**Comment by VaradK62442** [comment]
yep, ditto, good to merge

**Pull request merged**